aiohttp_devtools/runserver/serve.py

```diff
@@ -397,23 +397,49 @@ def _insert_footer(self, response: web.StreamResponse) -> web.StreamResponse:
         resp.last_modified = filepath.stat().st_mtime  # type: ignore[assignment]
         return resp
 
+    def _insert_footer_if_exists(
+        self, filename: str, response: web.StreamResponse
+    ) -> web.StreamResponse:
+        """Insert the footer if the file exists, otherwise return a 404 response."""
+        resolved_path = self._directory.joinpath(filename).resolve()
+        if not resolved_path.is_file():
+            return self._make_not_found_response(resolved_path)
+        return self._insert_footer(response)
+
+    def _make_not_found_response(self, raw_path: Path) -> web.StreamResponse:
+        """Create a 404 response with a list of available files under the requested path."""
+        while not raw_path.is_dir():
+            raw_path = raw_path.parent
+        paths = "\n".join(
+            " {}{}".format(p.relative_to(self._directory), "/" if p.is_dir() else "")
+            for p in raw_path.iterdir()
+        )
+        msg = "404: Not Found\n\nAvailable files under '{}/':\n{}\n".format(
+            raw_path.relative_to(self._directory), paths
+        )
+        return web.Response(text=msg, status=404, content_type="text/plain")
+
     async def _handle(self, request: web.Request) -> web.StreamResponse:
-        raw_path = self.modify_request(request)
+        loop = asyncio.get_running_loop()
+        raw_path = await loop.run_in_executor(None, self.modify_request, request)
```

Comment:
Probably a little overkill, given that it's literally a static server for 1 user. :P

Reply:
For sure, but if we don’t do it right, people will copy it, and it will end up in downstream applications that have a lot more users

Reply:
I think the risk of someone doing that is low for this code base, but I didn’t want to find out later that someone pointed out how it was done here as why it’s OK to do blocking IO in the event loop … because that’s never happened before to me